extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -309,55 +324,115 @@ func (a *genericActuator) deployMachineDeployments(ctx context.Context, cluster
+func (a *genericActuator) waitUntilWantedMachineDeploymentsAvailable(ctx context.Context, cluster *controller.Cluster, worker *extensionsv1alpha1.Worker, alreadyExistingMachineDeploymentNames sets.String, alreadyExistingMachineClassNames sets.String, wantedMachineDeployments extensionsworker.MachineDeployments, clusterAutoscalerUsed bool) error {
+	rollingUpdate := false
+	autoscalerIsScaledDown := false
+	workerStatusUpdatedForRollingUpdate := false
 
-		// Get the list of all existing machine deployments
-		existingMachineDeployments := &machinev1alpha1.MachineDeploymentList{}
-		if err := a.client.List(ctx, existingMachineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+	return retryutils.UntilTimeout(ctx, 5*time.Second, 3*time.Minute, func(ctx context.Context) (bool, error) {
+		var numHealthyDeployments, numUpdated, numAvailable, numUnavailable, numDesired, numberOfAwakeMachines int32
+
+		// Get the list of all machine deployments
+		machineDeployments := &machinev1alpha1.MachineDeploymentList{}
+		if err := a.client.List(ctx, machineDeployments, client.InNamespace(worker.Namespace)); err != nil {
+			return retryutils.SevereError(err)
+		}
+
+		// Get the list of all machine sets
+		machineSets := &machinev1alpha1.MachineSetList{}
+		if err := a.client.List(ctx, machineSets, client.InNamespace(worker.Namespace)); err != nil {
 			return retryutils.SevereError(err)
 		}
 
+		// map the owner reference to the machine sets
+		ownerReferenceToMachineSet := workerhelper.BuildOwnerToMachineSetMap(machineSets.Items)
+
 		// Collect the numbers of available and desired replicas.
-		for _, existingMachineDeployment := range existingMachineDeployments.Items {
+		for _, deployment := range machineDeployments.Items {
+			wantedDeployment := wantedMachineDeployments.GetDeploymentForName(deployment.Name)
+
 			// Filter out all machine deployments that are not desired (any more).
-			if !wantedMachineDeployments.HasDeployment(existingMachineDeployment.Name) {
+			if wantedDeployment == nil {
 				continue
 			}
 
+			machineSets := ownerReferenceToMachineSet[deployment.Name]
+
+			// use `wanted deployment` for these checks, as the existing deployments can be based on an outdated cache
+			alreadyExistingMachineDeployment := alreadyExistingMachineDeploymentNames.Has(wantedDeployment.Name)
+			newMachineClass := !alreadyExistingMachineClassNames.Has(wantedDeployment.ClassName)
+
+			if alreadyExistingMachineDeployment && newMachineClass {
+				rollingUpdate = true
+				a.logger.V(6).Info(fmt.Sprintf("Machine deployment %s is performing a rolling update", deployment.Name))
+				// Already existing machine deployments with a rolling update should have > 1 machine sets
+				if len(machineSets) <= 1 {
+					return retryutils.MinorError(errors.New(fmt.Sprintf("Waiting for the MachineControllerManager to create the machine sets for the machine deployment (%s/%s)...", deployment.Namespace, deployment.Name)))
+				}
```

Comment:
Why do we need this check? Doesn't it not matter at all?

Reply:
It is a sanity check.  I observed we need it when operating on a stale cache.
We want to avoid making the decision wether the rolling update already completed based on an outdated cache. This was being done before and lead to not waiting for the update to complete.

The good thing is that we know what the MCM should create, so we can check for these things and can easily detect abnormalities. So we can garuantee to make decisions based on current information.

Reply:
@prashanth26

Reply:
ok, I see, thanks.